Add eslint and prettier

wgwz opened the pull request:
## Description

Closes: https://github.com/regen-network/regen-registry/issues/756
Closes: https://github.com/regen-network/regen-registry/issues/862

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

This adds an eslint configuration that runs a few common rules. It also configures prettier to run via the eslint command. See the new additions to the README file for the commands that have been tested. We are also adding GH actions for CI that will run the linter and the tests for any non-master branches.

The most critical file to review is `.eslintrc.js`. Reviewing the output of `yarn run eslint .` locally will also be important. To see the GH action config see this file: `.github/workflows/registryServer.yml`.

This PR also fixed all of the formatting/linting issues that our eslint config raised.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] targeted `dev` branch
- [x] provided a link to the relevant issue or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all tests passed
- [ ] submit a PR against `master` branch after this one (and other PRs depending on this one, e.g. on regen-network/regen-web, if applicable) get merged

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**.*

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

blushi commented:
@wgwz is this r4r? (it's still set as a draft)

wgwz commented on .eslintrc.js:
```diff
@@ -0,0 +1,21 @@
+module.exports = {
+    "env": {
+        "es2021": true,
```
I'm not sure that we need this. I'm not too sure if we've decided to stick to a particular version of "es". I'm guessing we can probably just pull this out because it's not in regen-web.

blushi replied:
sounds good

wgwz commented on README.md:
````diff
@@ -114,6 +114,28 @@ yarn test
+To automatically apply fixes where possible use `--fix`:
+```sh
+yarn run eslint <path-to-file> --fix
+```
````
Should I add these commands to package.json to mirror regen-web? `lint` and `format`? Technically this set up I added does both things at once, so maybe it's not needed?

blushi commented on .eslintrc.js:
```diff
@@ -0,0 +1,20 @@
+    "plugins": [
+        "@typescript-eslint",
+    ],
+    "rules": {
+    }
```
let's use the same rules as in https://github.com/regen-network/regen-web/blob/master/.eslintrc.js or are they actually already inferred by the extends?

wgwz replied:
I believe that they are inferred by the extends. I recall reading that
during the process I followed to set this up.

blushi replied:
I tried to add/rm them locally and the error/warning list is actually a bit larger with these additional rules

wgwz replied:
ah, ok. i'll add these rules then, np. good to be consistent

blushi reviewed (commented):
Looks good so far, just a small question

wgwz commented:
I believe that they are inferred by the extends. I recall reading that 
during the process I followed to set this up.

On 3/8/22 04:38, Marie Gauthier wrote:
>
> ***@***.**** commented on this pull request.
>
> Looks good so far, just a small question
>
> ------------------------------------------------------------------------
>
> In .eslintrc.js 
> <https://github.com/regen-network/registry-server/pull/128#discussion_r821467500>:
>
> > +    "rules": {
> +    }
>
> let's use the same rules as in 
> https://github.com/regen-network/regen-web/blob/master/.eslintrc.js or 
> are they actually already inferred by the extends?
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/regen-network/registry-server/pull/128#pullrequestreview-902744136>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ACNGY4TUJAT4ACQGVQUKARTU64NX7ANCNFSM5P6VV6QA>.
> Triage notifications on the go with GitHub Mobile for iOS 
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> 
> or Android 
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

wgwz commented on .eslintrc.js:
```diff
@@ -0,0 +1,27 @@
+  plugins: ['@typescript-eslint'],
+  rules: {
+    'no-empty': 'off',
+    'no-console': 'warn',
```
@blushi should we turn this off? i feel that in a back end it's more acceptable/expected to have log statements.

blushi replied:
makes sense, although this was just a warning, it might make the code in the IDE less readable because of all the underlined console.log's

wgwz commented on server/routes/stripe.ts:
```diff
@@ -243,7 +256,7 @@ router.post(
           }
           break;
-        case 'checkout.session.completed':
+        case 'checkout.session.completed': {
```
i had to add this `{}` for this case clause to fix an eslint issue: https://eslint.org/docs/rules/no-case-declarations

i'm not sure how to test these changes, so i'm not 100% sure that this change will work.

blushi replied:
what do you mean by "how to test these changes"?

wgwz replied:
i just meant that i was unsure if this addition would break anything in stripe.js, and that i was looking for a way to confirm that it won’t break anything (by a manual test or a review :-)

wgwz commented on server/worker/tasks/send_email.ts:
```diff
@@ -1,6 +1,5 @@
 import * as AWS from 'aws-sdk';
 import * as nodemailer from 'nodemailer';
-// @ts-ignore
```
i removed this because eslint has an error that bans this

wgwz commented on server/worker/tasks/send_email.ts:
```diff
@@ -54,22 +52,22 @@ const task: Task = async (inPayload) => {
 
 const templatePromises = {};
-function loadTemplate(template: string) {
+function loadTemplate(template: string): Promise {
```
this is a pretty weak statement, and i'm not sure that it is even valid here.

blushi replied:
you can't just return a `Promise` like this:
![image](https://user-images.githubusercontent.com/3149358/157395920-d4d0215d-aaf5-44c9-b1fb-f27db6250209.png)
You might want to add a eslint plugin to your IDE if you don't have that already.

You need to add the corresponding returned value which is a function in this case:

```suggestion
function loadTemplate(template: string): Promise<(variables: { [varName: string]: any }) => Promise<string>> {
```

wgwz replied:
i was pretty sure this was not a valid return, really just wanted to get something here for review. i added types for other promises in our code, but this was one of the more complicated return values i came across. thanks for the help! i’ll try to find a plugin that helps with this.

wgwz commented:
@blushi i just wanted to point out that the easiest way to review the changes i've made in terms of the linting stuff might be to look at the commits tab, and then click through each one to see it's differences. at this point it's such a big set of changes it's hard to just look at the files changed tab. i took care to dedicate each commit to a specific adjustment i made based on the linter output.

as of my latest commit, this is the output i'm getting while running the linter (let me know what you think in terms of any of these errors that we have left to clean up):

```yarn run v1.22.15
$ /home/kyle/regen/registry-server/node_modules/.bin/eslint . --fix

/home/kyle/regen/registry-server/server/Server.ts
    2:13  warning  'path' is defined but never used                @typescript-eslint/no-unused-vars
   12:28  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   13:13  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   14:20  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   17:3   error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   82:25  warning  'user' is assigned a value but never used       @typescript-eslint/no-unused-vars
   94:9   error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   95:9   error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   96:9   error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   97:9   error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   98:9   error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   99:9   error    Require statement not part of import statement  @typescript-eslint/no-var-requires
  100:9   error    Require statement not part of import statement  @typescript-eslint/no-var-requires
  101:9   error    Require statement not part of import statement  @typescript-eslint/no-var-requires

/home/kyle/regen/registry-server/server/__tests__/db/functions/create_user_organization.test.ts
  10:7  warning  'updates' is assigned a value but never used  @typescript-eslint/no-unused-vars

/home/kyle/regen/registry-server/server/__tests__/db/functions/issue_credits.test.ts
   26:20  warning  Unexpected any. Specify a different type   @typescript-eslint/no-explicit-any
  641:40  warning  'party' is defined but never used          @typescript-eslint/no-unused-vars
  677:3   warning  'creditClassId' is defined but never used  @typescript-eslint/no-unused-vars

/home/kyle/regen/registry-server/server/__tests__/db/functions/transfer_credits.test.ts
  177:50  warning  'project' is assigned a value but never used  @typescript-eslint/no-unused-vars
  430:50  warning  'project' is assigned a value but never used  @typescript-eslint/no-unused-vars
  465:50  warning  'project' is assigned a value but never used  @typescript-eslint/no-unused-vars
  495:50  warning  'project' is assigned a value but never used  @typescript-eslint/no-unused-vars
  527:12  warning  Unexpected any. Specify a different type      @typescript-eslint/no-explicit-any

/home/kyle/regen/registry-server/server/__tests__/db/helpers.ts
   2:1   error    Require statement not part of import statement  @typescript-eslint/no-var-requires
  39:25  warning  Unexpected any. Specify a different type        @typescript-eslint/no-explicit-any
  67:69  warning  Unexpected any. Specify a different type        @typescript-eslint/no-explicit-any

/home/kyle/regen/registry-server/server/middleware/imageOptimizer.ts
  4:14  error  Require statement not part of import statement  @typescript-eslint/no-var-requires

/home/kyle/regen/registry-server/server/pool.ts
  1:16  warning  'Client' is defined but never used              @typescript-eslint/no-unused-vars
  6:3   error    Require statement not part of import statement  @typescript-eslint/no-var-requires

/home/kyle/regen/registry-server/server/routes/auth.ts
   5:20  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   9:26  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
  60:19  warning  'qres' is defined but never used                @typescript-eslint/no-unused-vars
  69:25  warning  'qres' is defined but never used                @typescript-eslint/no-unused-vars

/home/kyle/regen/registry-server/server/routes/buyers-info.ts
   5:27  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
  32:20  warning  'records' is defined but never used             @typescript-eslint/no-unused-vars

/home/kyle/regen/registry-server/server/routes/contact.ts
   5:27  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
  30:20  warning  'records' is defined but never used             @typescript-eslint/no-unused-vars

/home/kyle/regen/registry-server/server/routes/files.ts
  3:12  error  Require statement not part of import statement  @typescript-eslint/no-var-requires
  4:22  error  Require statement not part of import statement  @typescript-eslint/no-var-requires

/home/kyle/regen/registry-server/server/routes/mailerlite.ts
  4:20  error  Require statement not part of import statement  @typescript-eslint/no-var-requires

/home/kyle/regen/registry-server/server/routes/metadata-graph.ts
  3:20  error  Require statement not part of import statement  @typescript-eslint/no-var-requires

/home/kyle/regen/registry-server/server/routes/recaptcha.ts
  2:15  error  Require statement not part of import statement  @typescript-eslint/no-var-requires

/home/kyle/regen/registry-server/server/routes/stripe.ts
    3:13  warning  'jwks' is defined but never used                @typescript-eslint/no-unused-vars
    4:13  warning  'jwt' is defined but never used                 @typescript-eslint/no-unused-vars
    9:27  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   16:16  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
   17:20  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
  103:11  warning  'product' is assigned a value but never used    @typescript-eslint/no-unused-vars
  187:27  warning  'transfer' is assigned a value but never used   @typescript-eslint/no-unused-vars
  297:29  warning  'transfer' is assigned a value but never used   @typescript-eslint/no-unused-vars

/home/kyle/regen/registry-server/server/worker/dist/send_email.d.ts
  10:24  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

/home/kyle/regen/registry-server/server/worker/tasks/credits_transfer__send_confirmation.ts
   3:28  warning  'dateFormat' is defined but never used    @typescript-eslint/no-unused-vars
   3:40  warning  'numberFormat' is defined but never used  @typescript-eslint/no-unused-vars
  42:60  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

/home/kyle/regen/registry-server/server/worker/tasks/interest_buyers__send_confirmation.ts
   3:28  warning  'dateFormat' is defined but never used    @typescript-eslint/no-unused-vars
   3:40  warning  'numberFormat' is defined but never used  @typescript-eslint/no-unused-vars
  10:73  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

/home/kyle/regen/registry-server/server/worker/tasks/send_email.ts
  28:24  warning  Unexpected any. Specify a different type   @typescript-eslint/no-explicit-any
  33:50  warning  Unexpected any. Specify a different type   @typescript-eslint/no-explicit-any
  53:11  warning  'info' is assigned a value but never used  @typescript-eslint/no-unused-vars
  75:47  warning  Unexpected any. Specify a different type   @typescript-eslint/no-explicit-any

/home/kyle/regen/registry-server/server/worker/worker.ts
  1:17  error  Require statement not part of import statement  @typescript-eslint/no-var-requires
  3:8   error  Missing return type on function                 @typescript-eslint/explicit-function-return-type

✖ 63 problems (29 errors, 34 warnings)

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

blushi commented on server/routes/recaptcha.ts:
```diff
@@ -1,36 +1,47 @@
 const recaptchaV3Key = process.env.RECAPTCHA_V3_SECRET_KEY;
 
-function verify(request, response: express.Response, key: string) {
+function verify(request, response: express.Response, key: string): void {
```
this seems right to me

blushi commented:
The warnings "is defined but never used" can be removed easily by just removing the unused variables declarations.
I think we shouldn't keep any linting errors in the code since these will get flagged by CI anyway.

wgwz commented:
I'm having trouble figuring out why this patch is not working. This patch attempts to change the require import for the `pgPool` in `server/Server.ts`. I'm saying it's not working based on the output of `yarn serve` (which is shown below):

```
diff --git a/server/Server.ts b/server/Server.ts
index 071b8fa..2c140ef 100755
--- a/server/Server.ts
+++ b/server/Server.ts
@@ -14,7 +14,7 @@ import imageOptimizer from './middleware/imageOptimizer';
 // Per this issue: https://github.com/graphile-contrib/pg-many-to-many/issues/64
 import PgManyToManyPlugin from '@graphile-contrib/pg-many-to-many';
 import url from 'url';
-const { pgPool } = require('./pool');
+import pgPool from './pool';
 
 if (process.env.NODE_ENV !== 'production') {
   require('dotenv').config();
diff --git a/server/pool.ts b/server/pool.ts
index ce98765..384c3be 100644
--- a/server/pool.ts
+++ b/server/pool.ts
@@ -32,5 +32,6 @@ const runnerPromise = new Promise((resolve, reject) => {
     });
 });
 
+export default pgPool;
 exports.pgPool = pgPool;
 exports.runnerPromise = runnerPromise;
```

Output of `yarn serve`:

```
yarn run v1.22.15
$ ts-node Server.ts

/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:513
    return new TSError(diagnosticText, diagnosticCodes)
           ^
TSError: ⨯ Unable to compile TypeScript:
Server.ts(78,16): error TS2345: Argument of type 'Pool' is not assignable to parameter of type 'string | Pool | PoolConfig'.
  Type 'import("/home/kyle/regen/registry-server/node_modules/@types/pg/index").Pool' is not assignable to type 'import("/home/kyle/regen/registry-server/node_modules/pg-sql2/node_modules/@types/pg/index").Pool'.
    The types returned by 'connect()' are incompatible between these types.
      Type 'Promise<import("/home/kyle/regen/registry-server/node_modules/@types/pg/index").PoolClient>' is not assignable to type 'Promise<import("/home/kyle/regen/registry-server/node_modules/pg-sql2/node_modules/@types/pg/index").PoolClient>'.
        Type 'PoolClient' is missing the following properties from type 'PoolClient': _explainResults, startExplain, stopExplain

    at createTSError (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:513:12)
    at reportTSError (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:517:19)
    at getOutput (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:752:36)
    at Object.compile (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:968:32)
    at Module.m._compile (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:1056:42)
    at Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
    at Object.require.extensions.<computed> [as .ts] (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

It looks like this output is most helpful from that:

```
The types returned by 'connect()' are incompatible between these types.
Type 'Promise<import("/home/kyle/regen/registry-server/node_modules/@types/pg/index").PoolClient>'
is not assignable to type
'Promise<import("/home/kyle/regen/registry-server/node_modules/pg-sql2/node_modules/@types/pg/index").PoolClient>'.
```

It looks as though [the first argument to `postgraphile` here](https://github.com/regen-network/registry-server/blob/kyle/756-add-eslint/server/Server.ts#L78), is expecting a [`pg-sql2`](https://www.npmjs.com/package/pg-sql2) type for `PoolClient` whereas our code is just using a standard `pg` `PoolClient` type (??) Or maybe it's just that `@types/pg` dependency has two different versions in those paths?

Any recommendations on how to fix this? I'm assuming I'm probably doing something wrong, as I'm still quite new to the JS import syntax :-)

blushi commented:
@wgwz This seems related to our dependency on `@types/pg`, if you remove this dependency from server/package.json, this should work again
`postgraphile` already needs `@types/pg` as dep anyway so this might have caused conflicts.

wgwz commented:
> @wgwz This seems related to our dependency on `@types/pg`, if you remove this dependency from server/package.json, this should work again `postgraphile` already needs `@types/pg` as dep anyway so this might have caused conflicts.

Indeed! Thanks for the suggestion @blushi . This is now working via b220511

wgwz commented:
@blushi i just wanted to check that the changes i made in this patch are ok:

```
diff --git a/server/routes/files.ts b/server/routes/files.ts
index 92d8a24..5e7c517 100644
--- a/server/routes/files.ts
+++ b/server/routes/files.ts
@@ -1,7 +1,7 @@
 import * as express from 'express';
 import * as bodyParser from 'body-parser';
-const S3 = require('aws-sdk/clients/s3');
-const { Readable } = require('stream');
+import S3 from 'aws-sdk/clients/s3';
+import { Readable } from 'stream';
 const router = express.Router();
 
 const bucketName = process.env.AWS_S3_BUCKET;
@@ -30,7 +30,7 @@ router.post(
   (request, response: express.Response) => {
     const image = (request as FilesRequest).files.image;
     const key = request.body.filePath;
-    const fileStream = Readable.from(image.data);
+    const fileStream = Readable.from([image.data]);
```
the change above is the specific thing i wanted to check on.

without that change, and modifying the import for `Readable`, i'm getting this traceback:

```
kyle@lookfar4:~/regen/registry-server/server$ yarn start                                                                                                  
yarn run v1.22.15                                                                                                                                         
$ ts-node Server.ts                                                                                                                                       
[HPM] Proxy created: /  -> http://185.181.11.150:26657/                                                                                                   
[HPM] Proxy rewrite rule created: "^/ledger" ~> "/"                                                                                                       
[HPM] Proxy created: /  -> http://185.181.11.150:1317/                                                                                                    
[HPM] Proxy rewrite rule created: "^/ledger-rest" ~> "/"                                                                                                  
                                                                                                                                                          
/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:513                                                                             
    return new TSError(diagnosticText, diagnosticCodes)                                                                                                   
           ^                                                                                                                                              
TSError: ⨯ Unable to compile TypeScript:                                                                                                                  
routes/files.ts:33:38 - error TS2345: Argument of type 'File' is not assignable to parameter of type 'Iterable<any> | AsyncIterable<any>'.                
  Property '[Symbol.asyncIterator]' is missing in type 'File' but required in type 'AsyncIterable<any>'.                                                  
                                                                                                                                                          
33     const fileStream = Readable.from(image.data);                                                                                                      
                                        ~~~~~~~~~~                                                                                                        
                                                                                                                                                          
  ../node_modules/typescript/lib/lib.es2018.asynciterable.d.ts:40:5                                                                                       
    40     [Symbol.asyncIterator](): AsyncIterator<T>;                                                                                                    
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                    
    '[Symbol.asyncIterator]' is declared here.                                                                                                            
                                                                                                                                                          
    at createTSError (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:513:12)                                                   
    at reportTSError (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:517:19)                                                   
    at getOutput (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:752:36)
    at Object.compile (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:968:32)
    at Module.m._compile (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:1056:42)
    at Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
    at Object.require.extensions.<computed> [as .ts] (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

apparently `Readable` is asking for an iterable or an async iterable type, so that's why i just tried the naive thing of a making it an iterable. but it feels to easy to be right, and i trust that this code was working as it was. so yeah, i'm not quite sure what to do here, and i'm also not sure how i can test that change is really working (beyond typescript successfully compiling).

wgwz commented on server/routes/files.ts:
```diff
@@ -30,7 +30,7 @@ router.post(
     const image = (request as FilesRequest).files.image;
     const key = request.body.filePath;
-    const fileStream = Readable.from(image.data);
+    const fileStream = Readable.from([image.data]);
```
Hi @blushi here's the change that i need review on. the reason i added the brackets here was that it's an easy way to satisfy the type system. but i'm unsure if this would actually work at runtime. without the change i'm getting this traceback from the typescript compiler:

```
kyle@lookfar4:~/regen/registry-server/server$ yarn start                                                                                                  
yarn run v1.22.15                                                                                                                                         
$ ts-node Server.ts                                                                                                                                       
[HPM] Proxy created: /  -> http://185.181.11.150:26657/                                                                                                   
[HPM] Proxy rewrite rule created: "^/ledger" ~> "/"                                                                                                       
[HPM] Proxy created: /  -> http://185.181.11.150:1317/                                                                                                    
[HPM] Proxy rewrite rule created: "^/ledger-rest" ~> "/"                                                                                                  
                                                                                                                                                          
/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:513                                                                             
    return new TSError(diagnosticText, diagnosticCodes)                                                                                                   
           ^                                                                                                                                              
TSError: ⨯ Unable to compile TypeScript:                                                                                                                  
routes/files.ts:33:38 - error TS2345: Argument of type 'File' is not assignable to parameter of type 'Iterable<any> | AsyncIterable<any>'.                
  Property '[Symbol.asyncIterator]' is missing in type 'File' but required in type 'AsyncIterable<any>'.                                                  
                                                                                                                                                          
33     const fileStream = Readable.from(image.data);                                                                                                      
                                        ~~~~~~~~~~                                                                                                        
                                                                                                                                                          
  ../node_modules/typescript/lib/lib.es2018.asynciterable.d.ts:40:5                                                                                       
    40     [Symbol.asyncIterator](): AsyncIterator<T>;                                                                                                    
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                    
    '[Symbol.asyncIterator]' is declared here.                                                                                                            
                                                                                                                                                          
    at createTSError (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:513:12)                                                   
    at reportTSError (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:517:19)                                                   
    at getOutput (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:752:36)
    at Object.compile (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:968:32)
    at Module.m._compile (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:1056:42)
    at Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
    at Object.require.extensions.<computed> [as .ts] (/home/kyle/regen/registry-server/server/node_modules/ts-node/src/index.ts:1059:12)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

blushi replied:
I tested it locally, it's still working fine

wgwz commented:
@blushi i just wanted to provide a general update for this PR. i've officially fixed all of the errors i was getting from eslint, with our current config :partying_face:. so i think we are getting quite close. there are some warnings left from eslint though:

```
yarn run v1.22.15
$ /home/kyle/regen/registry-server/node_modules/.bin/eslint .

/home/kyle/regen/registry-server/server/__tests__/db/functions/issue_credits.test.ts
   26:20  warning  Unexpected any. Specify a different type   @typescript-eslint/no-explicit-any
  641:40  warning  'party' is defined but never used          @typescript-eslint/no-unused-vars
  677:3   warning  'creditClassId' is defined but never used  @typescript-eslint/no-unused-vars

/home/kyle/regen/registry-server/server/__tests__/db/helpers.ts
  40:25  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  68:69  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

/home/kyle/regen/registry-server/server/routes/auth.ts
  60:19  warning  'qres' is defined but never used  @typescript-eslint/no-unused-vars
  69:25  warning  'qres' is defined but never used  @typescript-eslint/no-unused-vars

/home/kyle/regen/registry-server/server/routes/buyers-info.ts
  32:20  warning  'records' is defined but never used  @typescript-eslint/no-unused-vars

/home/kyle/regen/registry-server/server/routes/contact.ts
  30:20  warning  'records' is defined but never used  @typescript-eslint/no-unused-vars

/home/kyle/regen/registry-server/server/routes/stripe.ts
  103:11  warning  'product' is assigned a value but never used   @typescript-eslint/no-unused-vars
  189:27  warning  'transfer' is assigned a value but never used  @typescript-eslint/no-unused-vars
  301:29  warning  'transfer' is assigned a value but never used  @typescript-eslint/no-unused-vars

/home/kyle/regen/registry-server/server/worker/tasks/credits_transfer__send_confirmation.ts
   3:28  warning  'dateFormat' is defined but never used    @typescript-eslint/no-unused-vars
   3:40  warning  'numberFormat' is defined but never used  @typescript-eslint/no-unused-vars
  42:60  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

/home/kyle/regen/registry-server/server/worker/tasks/interest_buyers__send_confirmation.ts
  10:73  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

/home/kyle/regen/registry-server/server/worker/tasks/send_email.ts
  28:24  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  33:50  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  64:45  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  77:47  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

✖ 20 problems (0 errors, 20 warnings)

Done in 3.63s.
```

there were some `no-unused-var` warnings that I did address but for the `no-unused-var` warnings that remain, i don't think these merit a change. I'm unsure whether or not we want to address the `no-explicit-any` warnings right now. some of these look fairly tricky to create a type for.

one other area i think i need feedback on is the fact that i added some `interface`s into a few locations the `server/__tests__`. one thing is that i'm sure they is a better way to organize those, if you have suggestions for that, it would be appreciated. the second thing is that maybe i duplicated types, from elsewhere in our code.
here's an example of where I added some interfaces: https://github.com/regen-network/registry-server/blob/kyle/756-add-eslint/server/__tests__/db/helpers.ts#L181-L186

blushi commented:
@wgwz I think we can still get rid of all these unused variables

wgwz commented on server/routes/stripe.ts:
```diff
@@ -98,22 +98,11 @@ router.post('/create-checkout-session', bodyParser.json(), async (req, res) => {
-        //     destination: product.metadata.account_id,
-        //   },
-        // },
         line_items: [
```
@blush just wanting to check these changes above. `product` was marked as unused, so i removed it. and since the comment was using it, i decided to remove the comment too.

blushi replied:
I'd prefer to comment `const priceObject = ...` and `const product = ...` declaration instead

wgwz replied:
Reverted that specific change, so these are now just commented out. i added back the commented `payment_intent_data` as well.

wgwz commented on .github/workflows/registryServer.yml:
```diff
@@ -0,0 +1,62 @@
+name: registry-server
```
i added a workflow for this that's intended to run on every branch that's not master. i tested the GH action locally using [act](https://github.com/nektos/act). it seems to be working as i've intended it to on my local. however i'm not sure how to get github to kick off a test run of this for us. i didn't create it through the "actions" tab. i just wrote in my local, and then pushed this up. is it that this code needs to exists in `master` for github to actually run it?

anyways, at the moment the intent is to run both lint/format checks and then tests, for any non-master branch. but that can change. maybe we want these checks to run on master as well. i wasn't sure what to pick.

blushi replied:
I think we should address tests in CI in a later PR because we're gonna need to setup some db for the tests to run which might need more work and is unrelated to the current PR

wgwz replied:
i also opened up the separate task for adding the tests to CI (even though it's done now, i think):

https://github.com/regen-network/regen-registry/issues/862

blushi commented:
@wgwz could you fix the conflicts?

wgwz commented:
> could you fix the conflicts?

@blushi done!

wgwz commented:
> could you also update the PR template: "confirmed all tests passed" => "confirmed all CI checks have passed" see in `.github` folder: `PULL_REQUEST_TEMPLATE.md` and `PULL_REQUEST_TEMPLATE/default.md`

@blushi done!

blushi reviewed (approved):
Great work, so glad to see linting and CI finally setup!

mhagel reviewed (approved):
Tested, looks good.
